x-pack/plugins/data_enhanced/server/search/session/check_running_sessions.ts

```diff
@@ -10,93 +10,198 @@ import {
-    );
-
-    if (!runningSearchSessionsResponse.total) return;
-
-    logger.debug(`Found ${runningSearchSessionsResponse.total} running sessions`);
-
-    const updatedSessions = new Array<SavedObjectsFindResult<SearchSessionSavedObjectAttributes>>();
-
-    let sessionUpdated = false;
-
-    await Promise.all(
-      runningSearchSessionsResponse.saved_objects.map(async (session) => {
-        // Check statuses of all running searches
-        await Promise.all(
-          Object.keys(session.attributes.idMapping).map(async (searchKey: string) => {
-            const updateSearchRequest = (
-              currentStatus: Pick<SearchSessionRequestInfo, 'status' | 'error'>
-            ) => {
-              sessionUpdated = true;
-              session.attributes.idMapping[searchKey] = {
-                ...session.attributes.idMapping[searchKey],
-                ...currentStatus,
-              };
-            };
-
-            const searchInfo = session.attributes.idMapping[searchKey];
-            if (searchInfo.status === SearchStatus.IN_PROGRESS) {
-              try {
-                const currentStatus = await getSearchStatus(client, searchInfo.id);
-
-                if (currentStatus.status !== SearchStatus.IN_PROGRESS) {
-                  updateSearchRequest(currentStatus);
+    await getAllSavedSearchSessions$(deps, config)
+      .pipe(
+        mergeMap(async (runningSearchSessionsResponse) => {
+          if (!runningSearchSessionsResponse.total) return;
+
+          logger.debug(`Found ${runningSearchSessionsResponse.total} running sessions`);
+
+          const updatedSessions = new Array<
+            SavedObjectsFindResult<SearchSessionSavedObjectAttributes>
+          >();
+
+          await Promise.all(
+            runningSearchSessionsResponse.saved_objects.map(async (session) => {
+              const updated = await updateSessionStatus(session, client, logger);
+              let deleted = false;
+
+              if (!session.attributes.persisted) {
+                if (isSessionStale(session, config, logger)) {
+                  deleted = true;
+                  // delete saved object to free up memory
+                  // TODO: there's a potential rare edge case of deleting an object and then receiving a new trackId for that same session!
+                  // Maybe we want to change state to deleted and cleanup later?
+                  logger.debug(`Deleting stale session | ${session.id}`);
+                  await savedObjectsClient.delete(SEARCH_SESSION_TYPE, session.id);
+
+                  // Send a delete request for each async search to ES
+                  Object.keys(session.attributes.idMapping).map(async (searchKey: string) => {
+                    const searchInfo = session.attributes.idMapping[searchKey];
+                    if (searchInfo.strategy === ENHANCED_ES_SEARCH_STRATEGY) {
+                      try {
+                        await client.asyncSearch.delete({ id: searchInfo.id });
```

Comment:
Also, I'm assuming this isn't working yet because of the PR we're still waiting on from ES. Are we sure it will still be this same method (`asyncSearch.delete`) or is there a different set of params/route we need to use?

Reply:
@jimczi will the API used for delete using admin permissions be the same, so we can use the existing ESClient?